internal/repositories/github/service.go

```diff
@@ -0,0 +1,136 @@
+	// database. The repo is identified by its name and project.
+	DeleteRepositoryByName(
+		ctx context.Context,
+		client ghclient.GitHubRepoClient,
+		projectID uuid.UUID,
+		repoOwner string,
+		repoName string,
+	) error
+	// DeleteRepositoryByID removes the webhook and deletes the repo from the
+	// database. The repo is identified by its database ID and project.
+	DeleteRepositoryByID(
+		ctx context.Context,
+		client ghclient.GitHubRepoClient,
+		projectID uuid.UUID,
+		repoID uuid.UUID,
+	) error
+}
+
+type repositoryService struct {
+	webhookManager webhooks.WebhookManager
+	store          db.Store
+	eventProducer  events.Interface
+}
+
+// NewRepositoryService creates an instance of the RepositoryService interface
+func NewRepositoryService(
+	webhookManager webhooks.WebhookManager,
+	store db.Store,
+	eventProducer events.Interface,
+) RepositoryService {
+	return &repositoryService{
+		webhookManager: webhookManager,
+		store:          store,
+		eventProducer:  eventProducer,
+	}
+}
+
+func (r *repositoryService) DeleteRepositoryByName(
+	ctx context.Context,
+	client ghclient.GitHubRepoClient,
+	projectID uuid.UUID,
+	repoOwner string,
+	repoName string,
+) error {
+	// assumption: provider name should always be `ghprovider.Github` for this Github-specific code
+	params := db.GetRepositoryByRepoNameParams{
+		Provider:  ghprovider.Github,
+		RepoOwner: repoOwner,
+		RepoName:  repoName,
+		ProjectID: projectID,
+	}
+	repo, err := r.store.GetRepositoryByRepoName(ctx, params)
+	if err != nil {
+		return err
+	}
+	return r.deleteRepository(ctx, client, &repo)
+}
+
+func (r *repositoryService) DeleteRepositoryByID(
+	ctx context.Context,
+	client ghclient.GitHubRepoClient,
+	projectID uuid.UUID,
+	repoID uuid.UUID,
+) error {
+	repo, err := r.store.GetRepositoryByIDAndProject(ctx, db.GetRepositoryByIDAndProjectParams{
+		ID:        repoID,
+		ProjectID: projectID,
+	})
+	if err != nil {
+		log.Printf("error is %v", err)
+		return err
+	}
+	return r.deleteRepository(ctx, client, &repo)
+}
+
+func (r *repositoryService) deleteRepository(ctx context.Context, client ghclient.GitHubRepoClient, repo *db.Repository) error {
+	var err error
+
+	// Cleanup any webhook we created for this project
+	// `webhooksManager.DeleteWebhook` is idempotent, so there are no issues
+	// re-running this if we have to retry the request.
+	webhookID := repo.WebhookID
+	if webhookID.Valid {
+		err = r.webhookManager.DeleteWebhook(ctx, client, repo.RepoOwner, repo.RepoName, webhookID.Int64)
+		if err != nil {
+			return fmt.Errorf("error creating webhook from: %w", err)
+		}
+	}
+
+	// then remove the entry in the DB
+	if err = r.store.DeleteRepository(ctx, repo.ID); err != nil {
+		return fmt.Errorf("error deleting repository from DB: %w", err)
+	}
+
+	// Telemetry logging
+	logger.BusinessRecord(ctx).Provider = repo.Provider
+	logger.BusinessRecord(ctx).Project = repo.ProjectID
+	logger.BusinessRecord(ctx).Repository = repo.ID
```

Comment:
Minor note -- it would be good to set these as early as possible, so that early exits still capture this telemetry.

I haven't made a big deal about it yet, but since you're cleaning things up, let's move this up to the point where we know `repo` (or even store the data that's in the proto, even if it's not normalized).

When we store this data only at the end of the operation, it means that we can't correlate e.g. 70 errors in the logs with the project that's having a bad day.